cips/cip-shwap_protocol.md

```diff
@@ -0,0 +1,378 @@
+the messages happens using externally provided data commitment.
+
+This creates a bunch of complexities with the [reference Golang implementation][gimpl] that are necessary if forking
+and substantially diverging the upstream is not an option. The naive question would be: "Why not make content
+verification after Bitswap provided it back over its API?" Intuitively, this would simplify much and would not require
+"hacking" CID. However, this has an important downside - the Bitswap, in such a case, would consider the request finalized
+and the content as fetched and valid, sending a DONT_WANT message to its peers. In contrast, the message might still be invalid
+according to the verification rules.
+
+However, Bitswap still requires multihashes and CID codecs to be registered. Therefore, we provide a table for the
+supported [share identifiers](#share-identifiers) with their respective multihash and CID codec codes. This table
+should be extended whenever any new share identifier is added.
+
+| Name     | Multihash | Codec  |
+|----------|-----------|--------|
+| RowID    | 0x7811    | 0x7810 |
+| SampleID | 0x7801    | 0x7800 |
+| DataID   | 0x7821    | 0x7820 |
+
+## Backwards Compatibility
+
+Swap is incompatible with the old sampling protocol.
+
+After rigorous investigation, the celestia-node team decided against _implementing_ backward compatibility with
+the old protocol into the node client due to the immense complications it brings. Instead, the simple and time-efficient
+strategy is transiently deploying infrastructure for old and new versions, allowing network participants to migrate
+gradually to the latest version. We will first deprecate the old version, and once the majority has migrated, we will
+terminate the old infrastructure.
+
+## Considerations
+
+### Security
+
+Shwap does not change the security model of Celestia's Data Availability network and changes the underlying
+protocol for data retrieval.
+
+Essentially, the network and its codebase get simplified and require less code and infrastructure to operate. This in turn
+decreases the amount of implementation vulnerabilities, DOS vectors, message amplification, and resource exhaustion attacks.
+However, new bugs may be introduced, as with any new protocol.
+
+### Protobuf Serialization
+
+Protobuf is a widely adopted serialization format and is used within Celestia's protocols. This was quite an obvious choice
+for consistency reasons, even though we could choose other more efficient and advanced formats like Cap'n Proto.
+
+### Sampling and Reconstruction
+
+Shwap deliberately avoids specifying sampling and reconstruction logic The sampling concerns on randomness selection and 
+sample picking are out of Shwap's scope and a matter of following CIPs. Shwap only provides messaging for sampling(via
+[SampleID](#sampleid) and [Sample container](#sample-container)). 
+
+## Reference Implementation
```

Comment:
Will there be a reference implementation for Shwap later added here?

Reply:
I linked below two implementations. Or do you think we should add the whole code here? That's going to be huge

Reply:
IMHO I think it's best to link out. we will see what @YazzyYaz thinks and then can move forward